src/ParquetMsg.chpl

```diff
@@ -705,6 +725,102 @@ module ParquetMsg {
     }
   }
 
+  inline proc getReaderIdx(fileNum: int, rgNum: int) {
+    // we can assume there won't be more than 1000 RGs in a file
+    return (fileNum*1000) + rgNum;
+  }
+
+  proc getRowGroupNums(ref distFiles, ref numRowGroups) {
+    coforall loc in distFiles.targetLocales() with (ref numRowGroups) do on loc {
+      var locFiles: [distFiles.localSubdomain()] string = distFiles[distFiles.localSubdomain()];
+      for i in locFiles.domain {
+        c_openFile(locFiles[i].localize().c_str(), getReaderIdx(i,0));
+        numRowGroups[i] = c_getNumRowGroups(getReaderIdx(i,0));
+        for j in 2..numRowGroups[i] {
+          c_openFile(locFiles[i].localize().c_str(), getReaderIdx(i,j-1));
+        }
+      }
+    }
+    var maxRowGroups = 0;
+    for val in numRowGroups do if maxRowGroups < val then maxRowGroups = val;
+    return maxRowGroups;
+  }
+  
+  proc fillSegmentsAndPersistData(ref distFiles, ref entrySeg, ref externalData, ref containsNulls, ref valsRead, dsetname, sizes, len, numRowGroups, ref bytesPerRG, ref startIdxs) throws {
+    var (subdoms, length) = getSubdomains(sizes);
+    coforall loc in distFiles.targetLocales() with (ref externalData, ref valsRead, ref bytesPerRG) do on loc {
+      var locFiles: [distFiles.localSubdomain()] string = distFiles[distFiles.localSubdomain()];
+      var locSubdoms = subdoms;
+
+      for i in locFiles.domain {
+        var fname = locFiles[i];
+        var locDsetname = dsetname;
+        for rg in 0..#numRowGroups[i] {
+          c_createRowGroupReader(rg, getReaderIdx(i,rg));
+          c_createColumnReader(locDsetname.localize().c_str(), getReaderIdx(i,rg));
+        }
+      }
+
+      var errs: [locFiles.domain] (bool, parquetErrorMsg) = [0..#locFiles.domain] (false, new parquetErrorMsg());
+      forall i in locFiles.domain {
+        var fname = locFiles[i];
+        var locDsetname = dsetname;
+        var startIdx = locSubdoms[i].low;
+        for rg in 0..#numRowGroups[i] {
+          var totalBytes = 0;
+          startIdxs[i][rg] = startIdx;
+
+          var numRead = 0;
+
+          if c_readParquetColumnChunks(fname.localize().c_str(), 8192, len, getReaderIdx(i,rg), c_ptrTo(numRead), c_ptrTo(externalData[i][rg]), c_ptrTo(containsNulls[i][rg]), c_ptrTo(errs[i][1].errMsg)) == ARROWERROR {
```

Comment:
why a batch size of 8192? I feel like I vaguely remember that empirically this had the best performance. if that is the case then I think adding a comment would be helpful

Reply:
Whoops, thanks for pointing this out, we actually have a private config const that we should be using here so that users can modify that value, I had hardcoded it in some testing, but we don't want that.